File: src/problems/p002.rs
//! Problem 2: Even Fibonacci numbers

solve!(expecting_answer: 4_613_732, with: || {
    struct Fib(u32, u32);

    impl Iterator for Fib {
        type Item = u32;

        fn next(&mut self) -> Option<u32> {
            let curr = self.0;
            self.0 = self.1;
            self.1 += curr;
            Some(curr)
        }
    }

    Fib(1, 2)
        .take_while(|&n| n < 4_000_000)
        .filter(|n| n % 2 == 0)
        .sum::<u32>() as u128
});

File: project_euler/p002.py
from . import solve


@solve(expected=4_613_732)
def p002():
    a, b = 1, 2
    fibs = [a, b]

    while b < 4_000_000:
        a, b = b, a + b
        fibs.append(b)

    return sum(n for n in fibs if n % 2 == 0)
